components/autocomplete/index.js

```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+	let n = node;
+	while ( n.lastChild ) {
+		n = n.lastChild;
+	}
+	return n;
+}
+
+/**
+ * Is the node a text node.
+ * @param {?Node} node the node to check.
+ * @returns {boolean} true if the node is a text node.
+ */
+function isTextNode( node ) {
+	return node !== null && node.nodeType === 3;
+}
+
+/**
+ * Return the node only if it is a text node, otherwise return null.
+ * @param {?Node} node the node to filter.
+ * @returns {?Node} the node or null if it is not a text node.
+ */
+function onlyTextNode( node ) {
+	return isTextNode( node ) ? node : null;
+}
+
+/**
+ * Find the index of the last charater in the text that is whitespace.
+ * @param {String} text the text to search.
+ * @returns {Number} the last index of a white space character in the text or -1.
+ */
+function lastIndexOfSpace( text ) {
+	for ( let i = text.length - 1; i >= 0; i-- ) {
+		if ( /\s/.test( text.charAt( i ) ) ) {
+			return i;
+		}
+	}
+	return -1;
+}
+
+export class Autocomplete extends Component {
 	static getInitialState() {
 		return {
-			isOpen: false,
-			search: /.*/,
+			search: /./,
 			selectedIndex: 0,
+			open: undefined,
+			query: undefined,
+			range: undefined,
 		};
 	}
 
 	constructor() {
 		super( ...arguments );
 
-		this.bindMenuNode = this.bindMenuNode.bind( this );
+		this.bindNode = this.bindNode.bind( this );
 		this.select = this.select.bind( this );
 		this.reset = this.reset.bind( this );
-		this.onBlur = this.onBlur.bind( this );
 		this.search = this.search.bind( this );
 		this.setSelectedIndex = this.setSelectedIndex.bind( this );
 
 		this.state = this.constructor.getInitialState();
 	}
 
-	bindMenuNode( node ) {
-		this.menuNode = node;
+	bindNode( node ) {
+		this.node = node;
+	}
+
+	replace( range, replacement ) {
+		const container = document.createElement( 'div' );
+		container.innerHTML = renderToString( replacement );
+		while ( container.firstChild ) {
+			const child = container.firstChild;
+			container.removeChild( child );
+			range.insertNode( child );
+			range.setStartAfter( child );
+		}
+		range.deleteContents();
 	}
 
 	select( option ) {
-		const { onSelect } = this.props;
+		const { open, range, query } = this.state;
+		const { onSelect } = open || {};
 
 		this.reset();
 
 		if ( onSelect ) {
-			onSelect( option );
+			const replacement = onSelect( option.value, range, query );
+			if ( replacement !== undefined ) {
+				this.replace( range, replacement );
+			}
 		}
 	}
 
 	reset() {
 		this.setState( this.constructor.getInitialState() );
 	}
 
-	onBlur( event ) {
-		// Check if related target is not within, in the case that the user is
-		// selecting an option by button click
-		if ( this.menuNode && ! this.menuNode.contains( event.relatedTarget ) ) {
-			this.reset();
+	// this method is separate so it can be overrided in tests
+	getCursor( container ) {
+		const selection = window.getSelection();
+		if ( selection.isCollapsed && container.contains( selection.anchorNode ) ) {
```

Comment:
Is there any chance we'd get to this point and `container.contains( selection.anchorNode )` be false? Seems it could be fair to assume `input` event would only fire if selection is within the container? Trying to see if we can drop some logic which is run on every keystroke in Editable.

Reply:
Ok, I have moved it into an assertion so it is only run during development.